File: scripts/clean_global_places.R
# Global places cleaning
# Usage:
#   Rscript scripts/clean_global_places.R [input_path] [--overwrite]
#
# Example:
#   Rscript scripts/clean_global_places.R data/intermediate/global/undated --overwrite

suppressPackageStartupMessages({
  library(jsonlite)
  library(purrr)
})

excluded_amenities <- c(
  "childcare",
  "school",
  "hospital",
  "social_facility",
  "college",
  "university",
  "kindergarten",
  "community_centre",
  "events_venue",
  "library",
  "pub",
  "grave_yard",
  "parking"
)

excluded_buildings <- c("school")

generic_name_tokens <- c(
  "and",
  "anglican",
  "apostolic",
  "assembly",
  "assemblies",
  "baptist",
  "catholic",
  "centre",
  "centres",
  "christian",
  "church",
  "churches",
  "community",
  "corps",
  "hall",
  "house",
  "houses",
  "methodist",
  "parish",
  "place",
  "presbyterian",
  "roman",
  "saint",
  "salvation",
  "site",
  "sites",
  "st",
  "the",
  "worship",
  "youth"
)

parse_args <- function(args) {
  positional <- args[!startsWith(args, "--")]
  flags <- args[startsWith(args, "--")]

  list(
    input_path = if (length(positional) >= 1) positional[[1]] else NULL,
    overwrite_outputs = any(flags %in% c("--overwrite", "--force"))
  )
}

find_latest_snapshot_dir <- function() {
  base_dir <- file.path("data", "intermediate", "global")
  if (!dir.exists(base_dir)) {
    stop("No intermediate global directory found at ", base_dir, ".")
  }

  snapshot_dirs <- list.dirs(base_dir, recursive = FALSE, full.names = TRUE)
  if (length(snapshot_dirs) == 0) {
    stop("No snapshot directories found under ", base_dir, ".")
  }

  snapshot_dirs[[which.max(basename(snapshot_dirs))]]
}

iter_input_files <- function(input_path) {
  if (file.exists(input_path) && !dir.exists(input_path)) {
    return(input_path)
  }

  sort(list.files(
    input_path,
    pattern = "_places_normalized\\.json$",
    full.names = TRUE
  ))
}

as_numeric_or_na <- function(value) {
  if (is.null(value) || identical(value, "") || is.na(value)) {
    return(NA_real_)
  }

  suppressWarnings(as.numeric(value))
}

haversine_distance_metres <- function(lat_1, lng_1, lat_2, lng_2) {
  lat_1_rad <- lat_1 * pi / 180
  lng_1_rad <- lng_1 * pi / 180
  lat_2_rad <- lat_2 * pi / 180
  lng_2_rad <- lng_2 * pi / 180

  delta_lat <- lat_2_rad - lat_1_rad
  delta_lng <- lng_2_rad - lng_1_rad

  haversine <- sin(delta_lat / 2) ^ 2 +
    cos(lat_1_rad) * cos(lat_2_rad) * sin(delta_lng / 2) ^ 2

  6371000 * 2 * asin(sqrt(haversine))
}

matches_pattern <- function(pattern, value) {
  grepl(pattern, value %||% "", perl = TRUE, ignore.case = TRUE)
}

normalise_name_tokens <- function(name) {
  normalised <- tolower(name %||% "")
  normalised <- gsub("\\bst\\s+", "saint ", normalised, perl = TRUE)
  tokens <- unlist(regmatches(normalised, gregexpr("[a-z]+", normalised, perl = TRUE)))
  setdiff(unique(tokens), generic_name_tokens)
}

is_support_building_duplicate <- function(record, records) {
  name <- record$name %||% ""

  hall_like_name <- matches_pattern("\\bhall\\b", name) &&
    !matches_pattern(
      "\\b(kingdom hall|gospel hall|mission hall|assembly hall|church of christ hall|christadelphian hall)\\b",
      name
    )

  centre_like_name <- matches_pattern("\\bcentre\\b", name) &&
    !matches_pattern(
      "\\b(church|christian|islamic|baha'i|bahai|bahá'í|worship|life|gospel|masjid|mosque|faith|revival|outreach|breakthrough|hope|buddhist|temple|celebration|family|heritage|mission|charity)\\b",
      name
    )

  house_like_name <- matches_pattern("\\bhouse\\b", name) &&
    !matches_pattern(
      "\\b(house of|church|worship|masjid|mosque|temple|faith|hope|grace|bread|breakthrough)\\b",
      name
    )

  if (
    !matches_pattern("\\b(church hall|parish centre|community centre)\\b", name) &&
      !hall_like_name &&
      !centre_like_name &&
      !house_like_name
  ) {
    return(FALSE)
  }

  lat <- as_numeric_or_na(record$lat)
  lng <- as_numeric_or_na(record$lng)

  if (is.na(lat) || is.na(lng)) {
    return(FALSE)
  }

  record_tokens <- normalise_name_tokens(name)
  denomination <- record$denomination %||% ""

  for (other in records) {
    if (identical(other$id, record$id)) {
      next
    }

    other_name <- other$name %||% ""
    if (!matches_pattern("\\b(church|cathedral|chapel|temple|mosque|synagogue|gurdwara)\\b", other_name)) {
      next
    }

    other_lat <- as_numeric_or_na(other$lat)
    other_lng <- as_numeric_or_na(other$lng)

    if (is.na(other_lat) || is.na(other_lng)) {
      next
    }

    if (haversine_distance_metres(lat, lng, other_lat, other_lng) > 100) {
      next
    }

    other_tokens <- normalise_name_tokens(other_name)
    shared_tokens <- intersect(record_tokens, other_tokens)
    denomination_match <- denomination != "" &&
      identical(denomination, other$denomination %||% "")

    if (length(shared_tokens) >= 1) {
      return(TRUE)
    }

    if (hall_like_name && denomination_match && length(record_tokens) <= 1) {
      return(TRUE)
    }

    if (centre_like_name && denomination_match && length(record_tokens) <= 1) {
      return(TRUE)
    }

    if (house_like_name && denomination_match && length(record_tokens) <= 1) {
      return(TRUE)
    }

    if (length(record_tokens) == 0 && denomination_match) {
      return(TRUE)
    }
  }

  FALSE
}

keep_record <- function(record, records) {
  tags <- record$tags_raw %||% list()
  amenity <- tags[["amenity"]] %||% NULL
  building <- tags[["building"]] %||% NULL
  name <- record$name %||% ""

  if (!is.null(amenity) && amenity %in% excluded_amenities) {
    return(FALSE)
  }

  if (!is.null(building) && building %in% excluded_buildings) {
    return(FALSE)
  }

  if (matches_pattern("\\b(cemetery|burial|urupa|office|residence|pub|kindergarten)\\b", name)) {
    return(FALSE)
  }

  if (matches_pattern("^Place of Worship \\d+$", name) && length(tags) == 0) {
    return(FALSE)
  }

  if (
    matches_pattern("^(Christian|Anglican|Roman_Catholic|Jewish|Sikh|Mormon|Methodist|Lutheran) Place of Worship$", name) &&
      is.null(amenity) &&
      is.null(building)
  ) {
    return(FALSE)
  }

  if (matches_pattern("\\bmasonic centre\\b", name)) {
    return(FALSE)
  }

  if (matches_pattern("\\bmasonic hall\\b", name)) {
    return(FALSE)
  }

  if (is_support_building_duplicate(record, records)) {
    return(FALSE)
  }

  if (matches_pattern("\\b(school|academy|seminary|college)\\b", name)) {
    chapel_like <- matches_pattern("chapel", name) || identical(building, "chapel")
    explicit_worship_space <- identical(amenity, "place_of_worship")

    if (!chapel_like && !explicit_worship_space) {
      return(FALSE)
    }
  }

  TRUE
}

clean_file <- function(path, output_dir, overwrite_outputs) {
  country_code <- toupper(sub("_places_normalized\\.json$", "", basename(path)))
  output_file <- file.path(
    output_dir,
    sub("_normalized", "_cleaned", basename(path))
  )

  if (file.exists(output_file) && !overwrite_outputs) {
    return(list(
      country_code = country_code,
      output_file = output_file,
      skipped = TRUE
    ))
  }

  message(sprintf("  [%s] Cleaning %s...", country_code, basename(path)))

  records <- read_json(path, simplifyVector = FALSE)
  filtered_records <- keep(
    records,
    \(record) keep_record(record, records)
  )

  write_json(
    filtered_records,
    output_file,
    pretty = TRUE,
    auto_unbox = TRUE,
    null = "null"
  )

  list(
    country_code = country_code,
    output_file = output_file,
    skipped = FALSE,
    input_count = length(records),
    output_count = length(filtered_records),
    removed_count = length(records) - length(filtered_records)
  )
}

write_cleaning_manifest <- function(results, output_dir, input_path) {
  manifest <- list(
    generated_at = format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC"),
    script = "scripts/clean_global_places.R",
    input_path = input_path,
    countries = results
  )

  manifest_path <- file.path(output_dir, "cleaning_manifest.json")
  write_json(manifest, manifest_path, pretty = TRUE, auto_unbox = TRUE, null = "null")
  message(sprintf("Wrote cleaning manifest: %s", manifest_path))
}

main <- function() {
  args <- parse_args(commandArgs(trailingOnly = TRUE))
  input_path <- args$input_path %||% find_latest_snapshot_dir()

  input_files <- iter_input_files(input_path)
  if (length(input_files) == 0) {
    stop("No normalized files found under ", input_path, ".")
  }

  snapshot_name <- if (dir.exists(input_path)) basename(input_path) else basename(dirname(input_path))
  output_dir <- file.path("data", "intermediate", "global", snapshot_name)
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

  results <- input_files |>
    map(\(path) clean_file(path, output_dir, args$overwrite_outputs))

  write_cleaning_manifest(results, output_dir, input_path)
}

main()

File: scripts/clean_global_places.py
#!/usr/bin/env python3
"""
Remove obvious non-worship records from normalized global country datasets.

This is intentionally conservative. It applies shared rule-based exclusions and
duplicate-support-building checks, while leaving ambiguous cases for review.
"""

from __future__ import annotations

import json
import math
import re
import sys
from datetime import UTC, datetime
from pathlib import Path


REPO_ROOT = Path("/Users/joseph/GIT/places-of-worship")
DEFAULT_INPUT_DIR = REPO_ROOT / "data" / "intermediate" / "global" / "undated"

EXCLUDED_AMENITIES = {
    "childcare",
    "school",
    "hospital",
    "social_facility",
    "college",
    "university",
    "kindergarten",
    "community_centre",
    "events_venue",
    "library",
    "pub",
    "grave_yard",
    "parking",
}
EXCLUDED_BUILDINGS = {"school"}
EXCLUDED_NAME_PATTERN = re.compile(
    r"\b(cemetery|burial|urupa|office|residence|pub|kindergarten)\b",
    re.IGNORECASE,
)
PLACEHOLDER_NAME_PATTERN = re.compile(r"^Place of Worship \d+$")
GENERIC_WORSHIP_LABEL_PATTERN = re.compile(
    r"^(Christian|Anglican|Roman_Catholic|Jewish|Sikh|Mormon|Methodist|Lutheran) Place of Worship$",
    re.IGNORECASE,
)
INSTITUTIONAL_NAME_PATTERN = re.compile(
    r"\b(school|academy|seminary|college)\b",
    re.IGNORECASE,
)
NON_WORSHIP_CENTRE_PATTERN = re.compile(
    r"\bmasonic centre\b",
    re.IGNORECASE,
)
NON_WORSHIP_HALL_PATTERN = re.compile(
    r"\bmasonic hall\b",
    re.IGNORECASE,
)
SUPPORT_BUILDING_PATTERN = re.compile(
    r"\b(church hall|parish centre|community centre)\b",
    re.IGNORECASE,
)
GENERIC_HALL_PATTERN = re.compile(r"\bhall\b", re.IGNORECASE)
GENERIC_CENTRE_PATTERN = re.compile(r"\bcentre\b", re.IGNORECASE)
GENERIC_HOUSE_PATTERN = re.compile(r"\bhouse\b", re.IGNORECASE)
GENUINE_HALL_NAME_PATTERN = re.compile(
    r"\b(kingdom hall|gospel hall|mission hall|assembly hall|church of christ hall|christadelphian hall)\b",
    re.IGNORECASE,
)
GENUINE_CENTRE_NAME_PATTERN = re.compile(
    r"\b(church|christian|islamic|baha'i|bahai|bahá'í|worship|life|gospel|masjid|mosque|faith|revival|outreach|breakthrough|hope|buddhist|temple|celebration|family|heritage|mission|charity)\b",
    re.IGNORECASE,
)
GENUINE_HOUSE_NAME_PATTERN = re.compile(
    r"\b(house of|church|worship|masjid|mosque|temple|faith|hope|grace|bread|breakthrough)\b",
    re.IGNORECASE,
)
PRIMARY_WORSHIP_NAME_PATTERN = re.compile(
    r"\b(church|cathedral|chapel|temple|mosque|synagogue|gurdwara)\b",
    re.IGNORECASE,
)
GENERIC_NAME_TOKENS = {
    "and",
    "anglican",
    "apostolic",
    "assembly",
    "assemblies",
    "baptist",
    "catholic",
    "centre",
    "centres",
    "christian",
    "church",
    "churches",
    "community",
    "corps",
    "hall",
    "house",
    "houses",
    "methodist",
    "parish",
    "place",
    "presbyterian",
    "roman",
    "saint",
    "salvation",
    "site",
    "sites",
    "st",
    "the",
    "worship",
    "youth",
}


def as_float(value: object) -> float | None:
    if value in (None, ""):
        return None

    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def haversine_distance_metres(
    lat_1: float,
    lng_1: float,
    lat_2: float,
    lng_2: float,
) -> float:
    lat_1_rad, lng_1_rad, lat_2_rad, lng_2_rad = map(
        math.radians,
        [lat_1, lng_1, lat_2, lng_2],
    )
    delta_lat = lat_2_rad - lat_1_rad
    delta_lng = lng_2_rad - lng_1_rad
    haversine = (
        math.sin(delta_lat / 2) ** 2
        + math.cos(lat_1_rad) * math.cos(lat_2_rad) * math.sin(delta_lng / 2) ** 2
    )
    return 6371000 * 2 * math.asin(math.sqrt(haversine))


def normalise_name_tokens(name: str) -> set[str]:
    normalised = name.lower().replace("st ", "saint ")
    tokens = set(re.findall(r"[a-z]+", normalised))
    return tokens - GENERIC_NAME_TOKENS


def is_support_building_duplicate(record: dict, records: list[dict]) -> bool:
    name = record.get("name") or ""
    hall_like_name = (
        GENERIC_HALL_PATTERN.search(name)
        and not GENUINE_HALL_NAME_PATTERN.search(name)
    )
    centre_like_name = (
        GENERIC_CENTRE_PATTERN.search(name)
        and not GENUINE_CENTRE_NAME_PATTERN.search(name)
    )
    house_like_name = (
        GENERIC_HOUSE_PATTERN.search(name)
        and not GENUINE_HOUSE_NAME_PATTERN.search(name)
    )

    if (
        not SUPPORT_BUILDING_PATTERN.search(name)
        and not hall_like_name
        and not centre_like_name
        and not house_like_name
    ):
        return False

    lat = as_float(record.get("lat"))
    lng = as_float(record.get("lng"))
    if lat is None or lng is None:
        return False

    record_tokens = normalise_name_tokens(name)
    denomination = record.get("denomination") or ""

    for other in records:
        if other.get("id") == record.get("id"):
            continue

        other_name = other.get("name") or ""
        if not PRIMARY_WORSHIP_NAME_PATTERN.search(other_name):
            continue

        other_lat = as_float(other.get("lat"))
        other_lng = as_float(other.get("lng"))
        if other_lat is None or other_lng is None:
            continue

        if haversine_distance_metres(lat, lng, other_lat, other_lng) > 100:
            continue

        other_tokens = normalise_name_tokens(other_name)
        shared_tokens = record_tokens & other_tokens
        denomination_match = (
            denomination != ""
            and denomination == (other.get("denomination") or "")
        )

        if len(shared_tokens) >= 1:
            return True

        if hall_like_name and denomination_match and len(record_tokens) <= 1:
            return True

        if centre_like_name and denomination_match and len(record_tokens) <= 1:
            return True

        if house_like_name and denomination_match and len(record_tokens) <= 1:
            return True

        if not record_tokens and denomination_match:
            return True

    return False


def keep_record(record: dict, records: list[dict]) -> bool:
    tags = record.get("tags_raw") or {}
    amenity = tags.get("amenity")
    building = tags.get("building")
    name = record.get("name") or ""

    if amenity in EXCLUDED_AMENITIES:
        return False

    if building in EXCLUDED_BUILDINGS:
        return False

    if EXCLUDED_NAME_PATTERN.search(name):
        return False

    if PLACEHOLDER_NAME_PATTERN.search(name) and not tags:
        return False

    if GENERIC_WORSHIP_LABEL_PATTERN.search(name) and amenity is None and building is None:
        return False

    if NON_WORSHIP_CENTRE_PATTERN.search(name):
        return False

    if NON_WORSHIP_HALL_PATTERN.search(name):
        return False

    if is_support_building_duplicate(record, records):
        return False

    if INSTITUTIONAL_NAME_PATTERN.search(name):
        chapel_like = "chapel" in name.lower() or building == "chapel"
        explicit_worship_space = amenity == "place_of_worship"
        if not chapel_like and not explicit_worship_space:
            return False

    return True


def iter_input_files(input_path: Path) -> list[Path]:
    if input_path.is_file():
        return [input_path]

    return sorted(input_path.glob("*_places_normalized.json"))


def latest_snapshot_dir() -> Path:
    base_dir = REPO_ROOT / "data" / "intermediate" / "global"
    snapshot_dirs = [path for path in base_dir.iterdir() if path.is_dir()]
    if not snapshot_dirs:
        raise FileNotFoundError(f"No snapshot directories found under {base_dir}")
    return sorted(snapshot_dirs)[-1]


def clean_file(input_file: Path, output_dir: Path, overwrite: bool) -> dict:
    country_code = input_file.name.replace("_places_normalized.json", "").upper()
    output_file = output_dir / input_file.name.replace("_normalized", "_cleaned")

    if output_file.exists() and not overwrite:
        return {
            "country_code": country_code,
            "output_file": str(output_file.relative_to(REPO_ROOT)),
            "skipped": True,
        }

    records = json.loads(input_file.read_text(encoding="utf-8"))
    filtered_records = [record for record in records if keep_record(record, records)]

    output_file.write_text(
        json.dumps(filtered_records, indent=2, ensure_ascii=False),
        encoding="utf-8",
    )

    return {
        "country_code": country_code,
        "output_file": str(output_file.relative_to(REPO_ROOT)),
        "skipped": False,
        "input_count": len(records),
        "output_count": len(filtered_records),
        "removed_count": len(records) - len(filtered_records),
    }


def main() -> None:
    args = sys.argv[1:]
    overwrite = "--overwrite" in args or "--force" in args
    positional = [arg for arg in args if not arg.startswith("--")]
    input_path = Path(positional[0]) if positional else latest_snapshot_dir()
    if not input_path.is_absolute():
        input_path = (REPO_ROOT / input_path).resolve()

    input_files = iter_input_files(input_path)
    if not input_files:
        raise FileNotFoundError(f"No normalized files found under {input_path}")

    snapshot_name = input_path.name if input_path.is_dir() else input_path.parent.name
    output_dir = REPO_ROOT / "data" / "intermediate" / "global" / snapshot_name
    output_dir.mkdir(parents=True, exist_ok=True)

    results = [clean_file(input_file, output_dir, overwrite) for input_file in input_files]

    manifest = {
        "generated_at": datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "script": "scripts/clean_global_places.py",
        "input_path": str(input_path.relative_to(REPO_ROOT)),
        "countries": results,
    }

    manifest_path = output_dir / "cleaning_manifest.json"
    manifest_path.write_text(json.dumps(manifest, indent=2), encoding="utf-8")
    print(f"Wrote cleaning manifest: {manifest_path}")


if __name__ == "__main__":
    main()
